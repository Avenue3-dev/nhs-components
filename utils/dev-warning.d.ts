/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export declare const useExperimentalWarning: (component: 'tabs' | 'character-count') => void;
/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export declare const useDevWarning: (message: string) => void;
