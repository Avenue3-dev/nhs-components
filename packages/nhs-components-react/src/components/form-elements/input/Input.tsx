import FormGroup from '@/components/miscellaneous/form-group/FormGroup'
import { FormElementProps } from '@/types/form-element-types'
import { InputWidth } from '@/types/nhsuk-sizes'
import clsx from 'clsx'
import {
  ForwardRefRenderFunction,
  HTMLInputTypeAttribute,
  HTMLProps,
  forwardRef,
} from 'react'

type InputProps = {
  width?: InputWidth
  type?: HTMLInputTypeAttribute
} & Omit<HTMLProps<HTMLInputElement>, 'type'> &
  FormElementProps

const render: ForwardRefRenderFunction<HTMLInputElement, InputProps> = (
  {
    id,
    width,
    className,
    label,
    labelProps,
    hint,
    hintProps,
    error,
    errorProps,
    disableErrorLine,
    type = 'text',
    ...rest
  },
  ref,
): JSX.Element => {
  return (
    <FormGroup
      id={id}
      label={label}
      labelProps={labelProps}
      hint={hint}
      hintProps={hintProps}
      error={error}
      errorProps={errorProps}
      disableErrorLine={disableErrorLine}
      inputType="input"
      render={({ elementId, hintId, labelId, name, hasError }) => (
        <input
          id={elementId}
          aria-describedby={hintId}
          aria-labelledby={labelId}
          name={name}
          className={clsx(
            'nhsuk-input',
            {
              [`nhsuk-input--width-${width}`]: width,
              'nhsuk-input--error': hasError,
            },
            className,
          )}
          ref={ref}
          type={type}
          {...rest}
        />
      )}
    />
  )
}

const Input = forwardRef(render)

export default Input
