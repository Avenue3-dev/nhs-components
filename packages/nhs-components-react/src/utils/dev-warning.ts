const prod = process.env.NODE_ENV === 'production'

export const experimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  if (prod) {
    return
  }

  console.warn(`ATTENTION
  This component is tagged as experimental and therefore could change without notice. See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}
