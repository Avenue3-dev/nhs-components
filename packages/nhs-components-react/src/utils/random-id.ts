export const generateRandomString = (length = 5): string => {
  const rand = Math.random() + 1

  return rand.toString(36).substring(2, length + 2)
}

export const generateRandomName = (prefix?: string): string => {
  const randString = generateRandomString()

  return prefix ? `${prefix}_${randString}` : randString
}

export const generateRandomId = generateRandomName
