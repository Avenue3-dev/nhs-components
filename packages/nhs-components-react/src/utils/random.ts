export const generateRandomString = (length: number = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
