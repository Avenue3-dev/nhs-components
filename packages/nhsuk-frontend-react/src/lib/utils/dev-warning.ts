const prod = process.env.NODE_ENV === 'production'

/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export const useExperimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  useDevWarning(`This component is tagged as experimental and therefore could change without notice.

  See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}

/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export const useDevWarning = (message: string): void => {
  if (prod) {
    return
  }

  console.warn(message)
}
