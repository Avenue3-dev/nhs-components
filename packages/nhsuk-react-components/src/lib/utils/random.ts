export const generateRandomString = (length = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
