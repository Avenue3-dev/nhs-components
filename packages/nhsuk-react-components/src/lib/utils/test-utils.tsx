/* eslint-disable import/export */
import {
  RenderOptions,
  RenderResult,
  cleanup,
  render,
} from '@testing-library/react'
import { MemoryRouter as Router } from 'react-router-dom'
import { afterEach } from 'vitest'

afterEach(() => {
  cleanup()
})

const customRender = (
  ui: React.ReactElement,
  options: RenderOptions = {},
): RenderResult => {
  return render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => <Router>{children}</Router>,
    ...options,
  })
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
