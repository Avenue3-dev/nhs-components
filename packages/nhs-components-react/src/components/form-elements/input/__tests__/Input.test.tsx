import { render, screen } from '@/utils/test-utils'
import { describe, expect, test } from 'vitest'
import Input from '../Input'

describe('Input', () => {
  test('should match the snapshot', () => {
    render(<Input data-testid="input" />)

    expect(screen.getByTestId('input').parentNode).toMatchSnapshot()
  })

  test('should match the snapshot - label', () => {
    render(<Input data-testid="input" label="First name" />)

    expect(screen.getByTestId('input').parentNode).toMatchSnapshot()
  })

  test('should match the snapshot - hint', () => {
    render(
      <Input data-testid="input" hint="Optionally provide your phone number" />,
    )

    expect(screen.getByTestId('input').parentNode).toMatchSnapshot()
  })

  test('should match the snapshot - error', () => {
    render(<Input data-testid="input" error="This is an error" />)

    expect(screen.getByTestId('input').parentNode).toMatchSnapshot()
  })
})
