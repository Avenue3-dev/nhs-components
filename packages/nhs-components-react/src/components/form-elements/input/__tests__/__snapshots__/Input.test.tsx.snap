// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > should match the snapshot - error 1`] = `
<div
  class="nhsuk-form-group nhsuk-form-group--error"
>
  <span
    class="nhsuk-error-message"
    for="input_:r3:"
    id="input_:r3:--error-message"
  >
    <span
      class="nhsuk-u-visually-hidden"
    >
      Error
      :
    </span>
    This is an error
  </span>
  <input
    class="nhsuk-input nhsuk-input--error"
    data-testid="input"
    id="input_:r3:"
    name="input_:r3:"
    type="text"
  />
</div>
`;

exports[`Input > should match the snapshot - hint 1`] = `
<div
  class="nhsuk-form-group"
>
  <div
    class="nhsuk-hint"
    for="input_:r2:"
    id="input_:r2:--hint"
  >
    Optionally provide your phone number
  </div>
  <input
    aria-describedby="input_:r2:--hint"
    class="nhsuk-input"
    data-testid="input"
    id="input_:r2:"
    name="input_:r2:"
    type="text"
  />
</div>
`;

exports[`Input > should match the snapshot - label 1`] = `
<div
  class="nhsuk-form-group"
>
  <label
    class="nhsuk-label"
    for="input_:r1:"
    id="input_:r1:--label"
  >
    First name
  </label>
  <input
    aria-labelledby="input_:r1:--label"
    class="nhsuk-input"
    data-testid="input"
    id="input_:r1:"
    name="input_:r1:"
    type="text"
  />
</div>
`;

exports[`Input > should match the snapshot 1`] = `
<div
  class="nhsuk-form-group"
>
  <input
    class="nhsuk-input"
    data-testid="input"
    id="input_:r0:"
    name="input_:r0:"
    type="text"
  />
</div>
`;
